{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is all the files that are important: \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The reference file with TR and TR id can be located at \n",
    "\n",
    "`/projectnb/vntrseek/NA12878/226782_HG38_rl100.seq`\n",
    "\n",
    "1st line:\n",
    "\n",
    "Repeatid,FirstIndex,LastIndex,CopyNumber,FastaHeader,FlankingLeft1000,Pattern,ArraySequence,FlankingRight1000,Conserved\n",
    "182169093,1197857,1197887,4.428571,chr1,TGTATCCATGAGGCTCACCCCCTGTGCAGCTAGGATGGAACTGCACCCCCTCCTCAGAGCAGCCCCCTCAGGCCAGAG CTTCAGACCCTCAGAGCTTCAGTGGACAAACAGGGGAGCAAGGCTATAGGAAGGAAGCAATCTCCCAGAGCCATCTGTCTGAATGAGGACCAGGGCCTCTGCCTCCTGCTGGCCCAGT\n",
    "GGGCTCGGGGTGAGGAGGGACTGACTGGCGTCTGGGGCAGGTATGGCTGCTGGAGATGAATTCTAACCCAGCCCTGCACACCAACTGCGAGGTCCTGAAGGAGGTCATCCCAGGTGTG\n",
    "GTCATCGAGACCCTGGGTGAGCCTCCAAGCCCCCACCCCACACCCCCACAACCTGCCCAGTCTTGAATGCCTACCTGCCCTCTTCCCAAGTTCAGACCCCCACAGATGGGGCTCTTCCACCTCCCGAGGGCCTGTGTGGCAGCGCCGCCCCCTGCCCCAACCTTCCCCACACCTCAGCTCACCTGGGTCCCACCCCTCACACCCTCCCCACCCCCTCCAGCCCCAGCCTCTGCCCC\n",
    "CCACTCACCCTCTCTCCCCCACCCGCACCAGACCTGGTGCTCGAGACCTTCCGGAAGAGCCTGCGCGGCCAGAAGATGTTGCCTCTGCTGTCCCAGCGCCGCTTCGTGCTCCTGCACA\n",
    "ACGGTGAGGCCGACCCGCGGCCGCACCTGGGGGGCTCGTGCAGCCTCCGCCGCTGGCCGCCCCTGCCCACCCGCCAGGCCAAGTCCTCCGGGCCACCCATGCCGCATGCCCCAGACCAGCCGGGCGCCCGCAGGCCTGCGCCACCTCCCTTGGTGCCGCAGCGTCCCCGGCCACCCGGCCCCGACCTGGACAGCGCCCACGATGGGGAGCCCCAGGCCCCGGGCACGGAGCAGTCG\n",
    "GGCACAGGCAACAGGCACCCGGCGCAAGAGCCTTCCCCGGGGACAGCCAAGGAGGAACGCGAGGAGCCTGAGAACGCGAGGCCCTAGGGG\n",
    "CAGCCA,CCCGCGC,CCCGCGCCCAGCGCCCCGCGCCCCGCGCCCC,AGCCGTGCTGCCTGCCCTCAGGGACCTATAAAGCCCACTTTGCTACAAACACAGTCTCTGCAGCAGAGTCCTGCCACCTCAGCGCCCCTGAGCCCATGCCCACCGGGGTCACCCCCACCACACCAGCACTGGCCTGAAAGATGCTCCCCTGGGACTTCCCCAGGCCCCCACTAGCCCCCAGTCCATGGCTGGGCCCTGTCCCTGGGGAGACTGTGCTGAGGGAGGGGGTCTCCGCGGCTGGGCCCAGTCCTGGGGGGCTGTTCTGAGGGAGGGGGTTCCCCACGGCTGGTCTATACTCCCTCCCGGGCTGCCCCATCCTGGGCTCCCGCCCACCACCCTGTGCCCCTCAGCACCTGGGCCCTCCTGCCGCCCTGTGACCCCACGGCCCCCATGGCCTGGCCGGTGATCCACCATTGCCTGTCATCCCCAGGAAGGGTCTGAACCCCTCTGGAGCTCCAAAGGAGGAGCCCCTTGAGCCCAGGATTGCTGCCGTGGGCACCCACAGAGCTCGCCCCAGCCTGCCTGGCACAGAACCTTCCTGGGCCTTTAAGCCCCAAGCCTGGCGCTCCAGCTCCCTGAGCCCCCAACCTTCCTGCAGTGGCCCCAGGACTTAGCCAGGGGTGAGGGGCACAGGGCAGCTAGCCCCAACAAGAGGGTCCCACAGAAGACATGCTAAGCCCACACTGGCCAGGCCCTGGTCAGCCCTGTGTAGCCGCTGATGGTTCTGTTCCCCATGGGTCTGGGGGTTTGCGAGGAGAGCTCTGGCGTCCCGGGGCCTCTGAGGCGTCCACACTACCCACCCCCTCACTCATCTCCCACCCAGAGAGCTAAAGTGTATGGGCCTCTTCCCAATGGCTGGCTCAAGGTGTCTGTGCCCCCCAAGTCGAGACCTGAGGATCGGGGGGCCCCGCCCCCAGGCCCACCCCCCACATGCCCTGCCTCCGACTCCACACGGTCACGGCCAAGTCGCCGGTCCCTCCTCCTGCCCTGCTCCGCAGCC,\n",
    "0.967742\n",
    "\n",
    "##### Modified this file to get desired fields to get this TR_reference file currently in `~/tr_reference_with_trid.bed`:\n",
    "\n",
    "Repeatid FastaHeader FirstIndex LastIndex CopyNumber\n",
    "\n",
    " ID   chr1 1197857 1197887 4.428571\n",
    "\n",
    "ID    chr1 1629847 1629871 3.571429\n",
    "\n",
    " ID   chr1 2435455 2435492 5.428571\n",
    "\n",
    " ID   chr1 3707809 3707839 4.428571\n",
    "    \n",
    " ID   chr1 4632534 4633461 133.000000\n",
    "\n",
    " ID   chr1 5509812 5509845 4.857143\n",
    "\n",
    " ID   chr1 5915203 5915232 4.285714\n",
    "\n",
    " ID   chr1 6613202 6613226 3.571429\n",
    "\n",
    " ID   chr1 6662119 6662149 4.571429\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Used this command to get this file: \n",
    "```mysql -h orca.bu.edu -u vntr -p -e\"SELECT map.refid AS reftrid, map.readid, fasta_reads.head AS header (fasta_ref_reps.firstindex-replnk.first)+1 AS readstart, fasta_ref_reps.lastindex+(LENGTH(fasta_reads.dna)-replnk.last) AS readstop FROM map INNER JOIN rank ON rank.refid=map.refid AND rank.readid=map.readid INNER JOIN rankflank ON rankflank.refid=map.refid AND rankflank.readid=map.readid INNER JOIN replnk ON replnk.rid=map.readid INNER JOIN fasta_reads on fasta_reads.sid=replnk.sid INNER JOIN fasta_ref_reps ON map.refid=fasta_ref_reps.rid WHERE map.bbb = 1 ORDER BY map.refid,map.readid\" VNTRPIPE_NA12878_10X > output1.txt```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The first 10 lines of `projectnb/vntrseek/NA12878/output1.txt`:\n",
    ">reftrid\treadid\theader\treadstart\treadstop \n",
    "\n",
    "\n",
    ">182168460\t8660395\t    ST-E00273:177:HMTTCCCXX:6:2117:24962:49724/2\t72065\t72208\n",
    "\n",
    ">182168460\t77629470\tST-E00273:177:HMTTCCCXX:4:2107:13748:23372/1\t72064\t72184\n",
    "\n",
    ">182168460\t77629471\tST-E00273:177:HMTTCCCXX:6:1107:22110:42675/1\t72063\t72183\n",
    "\n",
    ">182168460\t77629472\tST-E00273:177:HMTTCCCXX:4:2108:8724:8482/1\t72069\t72189\n",
    "\n",
    ">182168460\t77629473\tST-E00273:177:HMTTCCCXX:3:1101:21856:44275/1\t72070\t72190\n",
    "\n",
    ">182168460\t77629474\tST-E00273:177:HMTTCCCXX:7:1123:5680:6724/1\t72073\t72193\n",
    "\n",
    ">182168460\t77629475\tST-E00273:177:HMTTCCCXX:4:2116:21491:12314/1\t72065\t72185\n",
    "\n",
    ">182168460\t77629476\tST-E00273:177:HMTTCCCXX:3:2204:21653:16779/1\t72090\t72210\n",
    "\n",
    ">182168460\t77629477\tST-E00273:177:HMTTCCCXX:3:2219:7405:66003/1\t72076\t72196"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Here is the 10x file:\n",
    "##### located at projectnb/vntrseek/NA12878/bam_NA12878_GRCh38_phased_possorted_bam.bam\n",
    "\n",
    ">ST-E00273:177:HMTTCCCXX:6:2201:3488:50076\t147\tchr1\t9992\t27\t45S106M\t=\t10006\t-92\tCGCCCACCGCGCCCTACACTCTTTCCCTACACGACGCCCTTCCGATCTGTCCCATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCT\t<,(,7((7((<,,<,<F7<,<,,,FF<,,A,<7,<A,,,,FF<,,,<,,,FAA,<A7KKFAFFKKKFFAKKKFKFKKKAFAKKKFA7KKKKKFKKKKKAKKKKKFKKKKFAKKKKKKKKKKKKKKKKKKKKKKKFKKKKKAKKKKKFFFAA\tQT:Z:AAFFFKKK\tBC:Z:AGTGGAAC\tQX:Z:AAAFFFKKKKKKKKKK\tAM:A:0\tXM:A:0\tRX:Z:GTCCCATAACCCTAAC\tAS:f:-58\tRG:Z:27058:MissingLibrary:1:HMTTCCCXX:6\tXS:f:-60.5\tXT:i:1\tOM:i:13\n",
    "\n",
    ">ST-E00273:177:HMTTCCCXX:1:2112:10500:44503\t99\tchr1\t9993\t32\t6M1I94M27S\t=\t10001\t99\tCTGTCCTGGTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTACAGCGCTACAGCCCCCGCCCCAACCCC\tKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK7FF7FKAFKAKKKA7A,,,,,,,,,,A,,((,,(((7(,,,,,\tQT:Z:AAFFFKKK\tBC:Z:GTCTCCTT\tQX:Z:AAFFFKKKKKKKKKKK\tAM:A:0\tXM:A:0\tTR:Z:TCACTGA\tTQ:Z:KKKKKKK\tAS:f:-60.5\tRG:Z:27058:MissingLibrary:1:HMTTCCCXX:1\tXS:f:-64\tBX:Z:TATTACCAGGACAGTC-1\tXT:i:0\tRX:Z:TATTACCAGGACAGTC\tOM:i:15\n",
    "\n",
    ">ST-E00273:177:HMTTCCCXX:3:1202:11637:57794\t145\tchr1\t9993\t19\t87S64M\t=\t10256\t199\tCACCCCCCCCTCCCCCTTCTTCCTCCTACTCCCCCCCCCCAGATCTCCACCCCTACCCTTCCCCCCCCCCCTCCCCTCACCACTCGCCTGTCCCATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTA\t,(,,(((A<,(77,,,,,,,,,,,,,,,,((A((KFA7F<,,,,7,(,,7<A,77KAA,(<(F,((,<,<(<,A7,,,,,7,(7(,77,,,,,,,F7,KKKKK<KKKKKAKKKKKFKKKKKKKKKKKKKKKKKFKKKKKKKKKKKFFFFAA\tQT:Z:AAFFFKKK\tBC:Z:CAGATGGG\tQX:Z:AAFFFKKKKKKFKKKK\tAM:A:0\tXM:A:1\tRX:Z:TGAATCGTCGGTCTAA\tAS:f:-93.5\tRG:Z:27058:MissingLibrary:1:HMTTCCCXX:3\tXS:f:-95.5\tBX:Z:TGAATCGTCGGTCTAA-1\tXT:i:0\tOM:i:0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Made this file next with reed_header, barcode using 10x file:\n",
    "\n",
    "```samtools view bam_NA12878_GRCh38_phased_possorted_bam.bam | awk '{print $1 \"\\t\" substr($0,index($0,\"RX:\"),21)}' \n",
    "```\n",
    "\n",
    "Located at `projectnb/vntrseek/NA12878/read_tr.txt`\n",
    "\n",
    "first 10 lines: \n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:6:2201:3488:50076\tRX:Z:GTCCCATAACCCTAAC\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:3:1202:11637:57794\tRX:Z:TGAATCGTCGGTCTAA\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:2:1122:12165:66970\tRX:Z:TAGTGGTGTAGATGTA\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:1:2222:25063:43290\tRX:Z:CCATCGTGTGTCCGAC\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:3:2224:4300:31090\tRX:Z:GCTAAACGTAAACAGT\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:1:2216:15321:48072\tRX:Z:TCGATAGTCCGATACA\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:6:2211:26403:35414\tRX:Z:CAAGCGCTCGTCCGTT\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:4:2101:30107:72104\tRX:Z:ATCGGTAGTCTTACCC\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:8:1203:23673:32197\tRX:Z:TGACTTTTCCGAAGGA\n",
    "\n",
    "ST-E00273:177:HMTTCCCXX:8:1109:30502:66672\tRX:Z:ACCAAGTCATTCCGAA\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Loaded output1.txt from yozen and TR_ref.bed into a sqlite database and joined on TR_id:\n",
    "\n",
    " `select * from reedid_TR_reference join refTR on reftrid=Repeatid limit 10;`\n",
    " \n",
    " #### After loading this into the database and joining got this :\n",
    " \n",
    " TRID, Reedid, header, reed_start, reed_stop, tr_id, chr, firstindex, lastindex, subheader\n",
    " \n",
    " 182168460|8660395|ST-E00273:177:HMTTCCCXX:6:2117:24962:49724/2|72065|72208|182168460|chr1|72121|72164|6:2117:24962:49724/2\n",
    " \n",
    "182168460|77629470|ST-E00273:177:HMTTCCCXX:4:2107:13748:23372/1|72064|72184|182168460|chr1|72121|72164|4:2107:13748:23372/1\n",
    "\n",
    "182168460|77629471|ST-E00273:177:HMTTCCCXX:6:1107:22110:42675/1|72063|72183|182168460|chr1|72121|72164|6:1107:22110:42675/1\n",
    "\n",
    "182168460|77629472|ST-E00273:177:HMTTCCCXX:4:2108:8724:8482/1|72069|72189|182168460|chr1|72121|72164|4:2108:8724:8482/1\n",
    "\n",
    "182168460|77629473|ST-E00273:177:HMTTCCCXX:3:1101:21856:44275/1|72070|72190|182168460|chr1|72121|72164|3:1101:21856:44275/1\n",
    "\n",
    "182168460|77629474|ST-E00273:177:HMTTCCCXX:7:1123:5680:6724/1|72073|72193|182168460|chr1|72121|72164|7:1123:5680:6724/1\n",
    "\n",
    "182168460|77629475|ST-E00273:177:HMTTCCCXX:4:2116:21491:12314/1|72065|72185|182168460|chr1|72121|72164|4:2116:21491:12314/1\n",
    "\n",
    "182168460|77629476|ST-E00273:177:HMTTCCCXX:3:2204:21653:16779/1|72090|72210|182168460|chr1|72121|72164|3:2204:21653:16779/1\n",
    "\n",
    "182168460|77629477|ST-E00273:177:HMTTCCCXX:3:2219:7405:66003/1|72076|72196|182168460|chr1|72121|72164|3:2219:7405:66003/1\n",
    "\n",
    "182168460|77629478|ST-E00273:177:HMTTCCCXX:7:1105:16356:31898/1|72080|72200|182168460|chr1|72121|72164|7:1105:16356:31898/1\n",
    "\n",
    "###### added a subheader because thought it would be useful; it was not"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Adding the barcode to the read\n",
    "\n",
    "The sqlite database could not handle the 57g of data so I exported the file as a csv\n",
    "\n",
    "###### It is located here `/projectnb/vntrseek/NA12578/reed_sorted_ref.csv`\n",
    "\n",
    ">reftrid,readid,header,readstart,readstop,Repeatid,FastaHeader,FirstIndex,LastIndex,substrheader\n",
    "\n",
    ">183306238,121323100,ST-E00273:177:HMTTCCCXX:1:1101:10003:10873/2,29751056,29751206,183306238,chrX,29751090,29751129,1:1101:10003:10873/2\n",
    "\n",
    ">182792043,81459527,ST-E00273:177:HMTTCCCXX:1:1101:10003:9185/1,52231526,52231653,182792043,chr20,52231586,52231624,1:1101:10003:9185/1\n",
    "\n",
    ">182456503,49674388,ST-E00273:177:HMTTCCCXX:1:1101:10013:42780/1,103018285,103018412,182456503,chr13,103018295,103018336,1:1101:10013:42780/1\n",
    "\n",
    ">182495588,53232665,ST-E00273:177:HMTTCCCXX:1:1101:10023:68851/2,104189477,104189627,182495588,chr14,104189534,104189575,1:1101:10023:68851/2\n",
    "\n",
    ">182352089,41479770,ST-\n",
    "E00273:177:HMTTCCCXX:1:1101:10023:69309/1,88508050,88508177,182352089,chr11,88508089,88508116,1:1101:10023:69309/1\n",
    "\n",
    ">183092312,9531505,ST-E00273:177:HMTTCCCXX:1:1101:1002:17395/2,113688464,113688614,183092312,chr6,113688522,113688561,1:1101:1002:17395/2\n",
    "\n",
    ">182740416,57542640,ST-E00273:177:HMTTCCCXX:1:1101:1002:42358/1,164108743,164108870,182740416,chr2,164108777,164108812,1:1101:1002:42358/1\n",
    "\n",
    ">182592551,72000327,ST-E00273:177:HMTTCCCXX:1:1101:1002:58005/1,46185586,46185717,182592551,chr17,46185655,46185695,1:1101:1002:58005/1\n",
    "\n",
    ">182323575,43231033,ST-E00273:177:HMTTCCCXX:1:1101:10033:14934/2,13142343,13142493,182323575,chr11,13142355,13142384,1:1101:10033:14934/2\n",
    "\n",
    "###### Added the tag via linux join and sorted via tag:\n",
    "\n",
    "This file is located at `~/sorted_vntr_barcodes_reeds.csv(bed)`:\n",
    "\n",
    ">ST-E00273:177:HMTTCCCXX:3:2205:17594:37278,AAAAAAAAAAAAAAAA,182511914,55603980,57126660,57126810,182511914,chr15,57126697,57126730,3:2205:17594:37278/2\n",
    "ST-E00273:177:HMTTCCCXX:8:2222:21004:34395,AAAAAAAAAAAAAAAA,183287217,22152837,134144965,134145092,183287217,chr9,134145013,134145073,8:2222:21004:34395/1\n",
    "ST-E00273:177:HMTTCCCXX:7:2102:28158:4069,AAAAAAAAAAAAGATT,182880919,126292855,125273151,125273301,182880919,chr3,125273236,125273264,7:2102:28158:4069/2\n",
    "ST-E00273:177:HMTTCCCXX:7:2102:27539:4509,AAAAAAAAAAAAGATT,182880919,126292864,125273151,125273301,182880919,chr3,125273236,125273264,7:2102:27539:4509/2\n",
    "ST-E00273:177:HMTTCCCXX:8:1217:11018:67586,AAAAAAAAAATCAGGC,183084247,10120898,91146303,91146453,183084247,chr6,91146355,91146407,8:1217:11018:67586/2\n",
    "ST-E00273:177:HMTTCCCXX:2:1114:26545:17342,AAAAAAAAACAAAAAA,182672476,77945914,50395952,50396079,182672476,chr19,50396021,50396055,2:1114:26545:17342/1\n",
    "ST-E00273:177:HMTTCCCXX:5:2211:27143:20946,AAAAAAAAACGTTCAG,183330994,120075852,100631515,100631642,183330994,chrX,100631535,100631564,5:2211:27143:20946/1\n",
    "ST-E00273:177:HMTTCCCXX:5:2211:26717:18871,AAAAAAAAACGTTCAG,183330994,120075853,100631515,100631642,183330994,chrX,100631535,100631564,5:2211:26717:18871/1\n",
    "ST-E00273:177:HMTTCCCXX:4:1202:28909:69221,AAAAAAAAAGAAAGAA,182266704,29758920,8066044,8066194,182266704,chr10,8066057,8066085,4:1202:28909:69221/2\n",
    "ST-E00273:177:HMTTCCCXX:4:2223:31385:54383,AAAAAAAAAGGAAGAA,182575608,74627015,7428689,7428839,182575608,chr17,7428779,7428824,4:2223:31385:54383/2\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Marzie's file is located here sorted by tag: `~/barcodes_mapping.csv` and in bed format here `/projectnb/vntrseek/NA12878/barcodes_mapping_10x.bed`\n",
    ">AAAAAAAACGGAATTA,chr1,143211766,143255106,2,1238,43340,0.0285648\n",
    "AAAAAACGGAATTATC,chr1,125179052,125180517,5,1307,1465,0.8921502\n",
    "AAAAAACGGAATTATC,chr1,143185205,143232653,4,1069,47448,0.0225299\n",
    "AAAAAATTAGCCGGGC,chr1,64957678,65072162,4,488,114484,0.0042626\n",
    "AAAAACAGGAATAACT,chr1,143468825,144294909,2,914,826084,0.0011064\n",
    "AAAAACGGAATCATCG,chr1,143206557,143252044,3,961,45487,0.0211269\n",
    "AAAAACGGAATTATCG,chr1,143215788,143262478,5,1284,46690,0.0275005\n",
    "AAAAACTAGACAGAAT,chr1,123334577,124742801,7,2143,1408224,0.0015218\n",
    "AAAAACTAGACAGTAG,chr1,122323634,122473174,3,711,149540,0.0047546\n",
    "AAAAAGAGTGTTTCAA,chr1,122650973,123120432,4,967,469459,0.0020598"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now want to join the barcodes_mapping_10x file with sorted_vntr_barcodes_reeds file, intersect them and see if their barcodes are the same"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Intersect.sh contains the two final files `sorted_vntr_barcodes_reeds3.bed` and  `sorted_barcodes_mapping_10x.bed`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Intersecting the two files using pandas \n",
    "\n",
    "I made a program in python using pandas that successfully intersects the two files and also checks if the barcodes are the same. It is located at `~/Intersect_pandas.ipynb` and here is the code from that: \n",
    "\n",
    "```\n",
    "marzie_10x = pd.read_csv('test1intersect.csv', names = ['chr', 'start', 'end', 'tag', 'fragments','num1','num2','calculation'])\n",
    "vntr = pd.read_csv('test2intersect.csv', names = ['chr', 'reedstart','reedstop','header','tag','trid','reedid','trid2','firstindex','lastindex','subheader'])\n",
    "\n",
    "line  = marzie_10x.loc[marzie_10x['tag'] == 'AAAAACGGAATTATCG']\n",
    "line2 = vntr.iloc[1]\n",
    "\n",
    "intersected_10x = pd.DataFrame(columns=['tag', 'chr', 'marzie_start', 'marzie_stop', 'reedstart', 'reedstop','trid'])\n",
    "line_count = 0\n",
    "\n",
    "\n",
    "count = 0\n",
    "for x in range(len(vntr.header)): # for the number of reeds in vntr_file\n",
    "    tag1 = vntr.tag[x] # get their tags\n",
    "    line_vntr = vntr.iloc[x] # get that associated line\n",
    "    line1_marzie = marzie_10x.loc[marzie_10x['tag'] == str(tag1)] # check if marzie's file has the vntr tag\n",
    "\n",
    "    if line1_marzie.empty == False: # if it does\n",
    "        matching_chromosomes = line1_marzie\n",
    "        reedstart_vntr = vntr.iloc[x].reedstart\n",
    "        reedstop_vntr = vntr.iloc[x].reedstop\n",
    "        for y in range(len(matching_chromosomes.tag)): # need to iterate over all the matching tags that show up\n",
    "            reed_start_marzie = matching_chromosomes.iloc[y].start\n",
    "            reed_stop_marzie = matching_chromosomes.iloc[y].end\n",
    "            matching_line_marzie= line1_marzie.iloc[y]\n",
    "            if int(reedstart_vntr) in range(int(reed_start_marzie),int(reed_stop_marzie)): # check to see if the vntr range is in the marzie range \n",
    "                intersected_10x.loc[line_count] = [matching_line_marzie.tag,matching_line_marzie.chr,reed_start_marzie,reed_stop_marzie,[int(line_vntr.reedstart)][0],[int(line_vntr.reedstop)][0],[int(line_vntr.trid)][0]]\n",
    "                line_count = line_count + 1\n",
    "                \n",
    "intersected_10x.to_csv('intersect_test3.csv')\n",
    "``` \n",
    "\n",
    "This program writes a csv called intersect_test3.csv also located in the home directory for the files, \n",
    "\n",
    "Tested against bedtools intersect -a file -b file and the output was the same for 2 small bed files, also made all the barcodes the same and saw what happened. So it was a success, now I just need to run it on the two big files and hopefully get the same number as marzie. \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wrote a shell script located at `~/intersect_python.sh ` which runs the `intersect_script.py`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The python script should give me a file of all the ones that were mapped correctly by 10x genomics. That file is located at  `~/intersected_10x_vntr_with_chr2.csv` it wil contain\n",
    "\n",
    ">tag\tchr\tmarzie_start\tmarzie_stop\treedstart\treedstop\ttrid\treed_header\tvntr_chr\n",
    "\n",
    "> 0\tAAAAACGGAATTATCG\t1\t143215788.0\t143262478.0\t143220494.0\t143220644.0\t182219147.0\tST-\n",
    "E00273:177:HMTTCCCXX:6:1202:5172:16815\t1\n",
    "\n",
    "This file is all the reeds that came from 10x mapped through vntrseek and then joined on the TR id identified by vntrseek to get their reed_id, then matched with their tags from 10x genomics associated with that reed, and then compared this tag associated reed file with the file provided to me by marzie. \n",
    "\n",
    "Now need to check how many unmatched TR's were identified as indistinguishable by vntr for the reference file. \n",
    "\n",
    "Need to know how many tags span a given tr, how many regions a tag covers, how long are the regions that the tag covers \n",
    ">tags span tr = `/projectnb/vntrseek/NA12878/intersect_pandas_rahul/parallel_something/tags_span_tr.csv`\n",
    ">regions_tag_cover = `/projectnb/vntrseek/NA12878/span_of_tags_10x.bed`\n",
    "\n",
    ">length of tag covering regions = `isnt it just 150 bp`\n",
    "\n",
    "Want to know also how many unique tr's were mapped correctly to 10x and how many were not also. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently have four jobs running, \n",
    "\n",
    "1) `intersect_10x_vntr_faster3` sorts by tags but is wrong because it does not filter by chr \n",
    "2) `intersect6` sorts by tags and chromosomes and fileters by chromosomes too, so this one is correct\n",
    "3) `intersect7` is the same as 6 but in the correct directory \n",
    "3) `rahul_intersect_dynamic` (the first one) writes to my home directory so will not be able to write it because my home directory can't take more than 10 gigs\n",
    "4) `rahul_intersect_dynamic2` will write 3 files \n",
    "> intersected_dynamic_2.csv \n",
    "\n",
    ">intersected_failed_dynamic_2.csv\n",
    "\n",
    ">no_tag_found_intersect_dynamic_2.csv\n",
    "\n",
    "They are located here `/projectnb/vntrseek/NA12728/intersect_pandas_dynamic`\n",
    "and this last one will include the failures too and will filter by chr and tag, intersected_failed will capture the vntrseek reeds that did not map correctly and no_tag_found will cover any reed that does not have a matching tag in marzie's file. Should also be faster because only opening and writing to file not maintaning a database in the system. Although it does maintain the thing in memory, so maybe theres a better way to do this. \n",
    "\n",
    "`intersect_10x_vntr_faster` died and was aborted because it took too much time\n",
    "\n",
    "`intersect6` was also aborted because it took too much time \n",
    "`intersect7` was also aborted, was in the correct directory\n",
    "\n",
    "\n",
    "submitted new job using intel distribution of python which speeds up python by a lot\n",
    "\n",
    "`intel_python_faster_something` will produce 2 files:\n",
    "> intersected_dynamic3 = the actual intersected ones \n",
    "\n",
    ">intersected_failed_dynamic3.csv = only the failed ones inside the loop\n",
    "\n",
    ">no_tag_found_intersect_dynamic3.csv = this file will be empty \n",
    "\n",
    "`parallel python intel` will produce three files hopefully faster "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok so the file with how many tags each tr has is here: `\\projectnb\\vntrseek\\NA12878\\intersect_pandas_rahul\\tags_span_tr_vntr.csv`\n",
    "\n",
    "Example: \n",
    ">trid,number_of_unique_barcodes\n",
    "\n",
    ">183084247,13\n",
    "\n",
    ">18182672476,22\n",
    "\n",
    ">183330994,26\n",
    "\n",
    ">182266704,25\n",
    "\n",
    ">182575608,17\n",
    "\n",
    ">182172155,20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Need to speed up my code somehow for the original intersect_pandas functionality "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is current code: \n",
    "\n",
    "``` #importing pandas\n",
    "import csv \n",
    "import pandas as pd \n",
    "import multiprocessing as mp\n",
    "\n",
    "#setting the parameters for the two csv's \n",
    "names_vntr = ['header', 'tag', 'trid', 'reedid', 'reedstart', 'reedstop','trid2','chr', 'firstindex', 'lastindex', 'subheader']\n",
    "names_marzie = ['tag', 'chr', 'start', 'end', 'fragments', 'num1', 'num2', 'calulation']\n",
    "dtypes_marzie  = {'tag': 'str', 'chr':'str', 'start':'int', 'end':'int', 'fragments':'int', 'num1':'int', 'num2':'int', 'calulation':'float'}\n",
    "dtypes_vntr = {'header':'str', 'tag':'str', 'trid':'int','reedid':'int', 'reedstart':'int', 'reedstop':'int','trid2':'int','chr':'str', 'firstindex':'int', 'lastindex':'int', 'subheader':'str'}\n",
    "\n",
    "#loading the two csv's\n",
    "marzie_10x = pd.read_csv('barcodes_mapping_witho_chr.csv',names = names_marzie, dtype = dtypes_marzie)\n",
    "vntr = pd.read_csv('sorted_without_chr_vntr.csv',names = names_vntr, dtype = dtypes_vntr)\n",
    "\n",
    "\n",
    "#can read a line at a time\n",
    "#test1 = pd.read_csv('untitled.csv', names = ['name','num'], dtype = {'name': 'str', 'num':'int'})\n",
    "\n",
    "#sort by tags and chr \n",
    "sorted_vntr = vntr.sort_values(by= ['tag','chr'])\n",
    "sorted_marzie = marzie_10x.sort_values(by = ['tag', 'chr'])\n",
    "\n",
    "\n",
    "#setting another dataframe for the output\n",
    "\n",
    "\n",
    "#opening the necessary files \n",
    "intersected_dynamic = open('intersected_dynamic_test2.csv', 'a') \n",
    "intersected_failed = open('intersected_failed_dynamic_test2.csv', 'a')\n",
    "no_tag_found = open('no_tag_found_intersect_dynamic_test2.csv', 'a')\n",
    "\n",
    "intersected_dynamic.write('tag,chr,marzie_start,marzie_stop,reedstart,reedstop,trid,reed_header,vntr_chr' + \"\\n\")\n",
    "intersected_failed.write('header,tag,trid,reedid,reedstart,reedstop,vntr_chr,firstindex,lastindex' + \"\\n\")\n",
    "no_tag_found.write('header,tag,trid,reedid,reedstart,reedstop,vntr_chr,firstindex,lastindex' + \"\\n\")\n",
    "\n",
    "\n",
    "#function \n",
    "\n",
    "for x in range(len(sorted_vntr.header)): # for the number of reeds in vntr_file\n",
    "    tag1 = sorted_vntr.tag[x] # get their tags\n",
    "    line_vntr = sorted_vntr.iloc[x] # get that associated line\n",
    "    line1_marzie = sorted_marzie.loc[sorted_marzie['tag'] == str(tag1)] # check if marzie's file has the vntr tag\n",
    "\n",
    "    if line1_marzie.empty == False: # if it does\n",
    "        matching_chromosomes = line1_marzie\n",
    "        reedstart_vntr = sorted_vntr.iloc[x].reedstart\n",
    "        reedstop_vntr = sorted_vntr.iloc[x].reedstop\n",
    "        num_of_tags = len(matching_chromosomes.tag) #number of tags (if there are 5 tags this will be 5)\n",
    "        mismatched_marzie_reeds= 0\n",
    "        for y in range(num_of_tags): # need to iterate over all the matching tags that show up \"[0,1,2,3,4]\"\n",
    "            reed_start_marzie = matching_chromosomes.iloc[y].start\n",
    "            reed_stop_marzie = matching_chromosomes.iloc[y].end\n",
    "            matching_line_marzie= line1_marzie.iloc[y]\n",
    "            if matching_line_marzie.chr == line_vntr.chr and int(reedstart_vntr) in range(int(reed_start_marzie),int(reed_stop_marzie)): # check to see if the vntr range is in the marzie range and have the same chr\n",
    "\n",
    "                intersected_dynamic.write(str(matching_line_marzie.tag)+','+str(matching_line_marzie.chr) + ',' + str(reed_start_marzie)+ ',' +str(reed_stop_marzie) + ',' +str(line_vntr.reedstart) + ',' +str(line_vntr.reedstop) + ',' +str(line_vntr.trid)+ ',' +str(line_vntr.header)+ ',' +str(line_vntr.chr)+\"\\n\")\n",
    "                # if it is in the range and has the same chromosome then add to the intersected data frame\n",
    "            else:\n",
    "                mismatched_marzie_reeds = mismatched_marzie_reeds + 1 # will count how many mismatched_tags have been processed so far\n",
    "                if mismatched_marzie_reeds == num_of_tags: # if you cycle through all the tags\n",
    "\n",
    "                    intersected_failed.write(str(line_vntr.header)+','+str(line_vntr.tag) + ',' + str(line_vntr.trid)+ ',' +str(line_vntr.reedid) + ',' +str(line_vntr.reedstart) + ',' +str(line_vntr.reedstop) + ',' +str(line_vntr.chr)+ ',' +str(line_vntr.firstindex)+ ',' +str(line_vntr.lastindex)+\"\\n\")\n",
    "\n",
    "                #else if it has cycled through all the matched tags and has not found a match add that vntr reed to a failed dataframe\n",
    "    else: # it it didnt even find any tags add it to the failed data frame\n",
    "\n",
    "        no_tag_found.write(str(line_vntr.header)+','+str(line_vntr.tag) + ',' + str(line_vntr.trid)+ ',' +str(line_vntr.reedid) + ',' +str(line_vntr.reedstart) + ',' +str(line_vntr.reedstop) + ',' +str(line_vntr.chr)+ ',' +str(line_vntr.firstindex)+ ',' +str(line_vntr.lastindex)+\"\\n\")\n",
    "\n",
    "        \n",
    "intersected_dynamic.close()\n",
    "intersected_failed.close()\n",
    "no_tag_found.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have another test file that made the thing multiprocessing, in this folder `projectnb/vntrseek/NA12878/intersect_pandas_rahul/parallel_processing_intersect/`\n",
    "\n",
    "Intermediate files vntr: have intermediate file of the process"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graphs "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "first file : `tags_span_tr_vntr.csv` \n",
    "> trid,number_of_unique_barcode\n",
    "\n",
    ">182511914,23\n",
    ">183287217,8, \n",
    ">182880919,33\n",
    ">183084247,18\n",
    ">182672476,22\n",
    ">183330994,26\n",
    ">182266704,25\n",
    ">182575608,17\n",
    ">182172155,20\n",
    ">182826800,22\n",
    ">182629572,33\n",
    ">182713236,308\n",
    "\n",
    "first 12 lines "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " second file: in process right now  `tag_regions_marzie.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
